test/CrateTest/PDO/Http/ServerPoolTest.php

```diff
@@ -0,0 +1,238 @@
+                [
+                    'base_uri'           => 'https://localhost:4200',
+                    RequestOptions::CERT => 'foo.pem',
+                ],
+            ],
+            [
+                [
+                    PDO::CRATE_ATTR_SSL_MODE          => PDO::CRATE_ATTR_SSL_MODE_REQUIRED,
+                    PDO::CRATE_ATTR_SSL_CERT          => 'foo.pem',
+                    PDO::CRATE_ATTR_SSL_CERT_PASSWORD => 'foo',
+                ],
+                [
+                    'base_uri'           => 'https://localhost:4200',
+                    RequestOptions::CERT => ['foo.pem', 'foo'],
+                ],
+            ],
+        ];
+    }
+
+    /**
+     * @dataProvider pdoOptions
+     *
+     * @param array $options
+     * @param array $expected
+     */
+    public function testGuzzleClientOptionTest(array $options, array $expected)
+    {
+        foreach ($options as $attr => $val) {
+            $this->pdo->setAttribute($attr, $val);
+        }
+
+        $expectedWithDefaults = array_merge([
+            RequestOptions::TIMEOUT         => 0.0,
+            RequestOptions::CONNECT_TIMEOUT => 0.0,
+            RequestOptions::JSON            => [
+                'stmt' => 'query',
+                'args' => [],
+            ],
+
+            RequestOptions::HEADERS => [
+                'Default-Schema' => 'doc',
+            ],
+
+            RequestOptions::AUTH => null,
+            'base_uri'           => 'http://localhost:4200',
+        ], $expected);
+
+        $body = json_encode(['rows' => [], 'cols' => [], 'duration' => 0, 'rowcount' => 0]);
+
+        $this->client
+            ->expects($this->once())
+            ->method('request')
+            ->with('POST', '/_sql', $expectedWithDefaults)
+            ->willReturn(new Response(200, [], $body));
+
+        $this->serverPool->execute('query', []);
+    }
+
+    public function testExecuteWithNoRespondingServers()
+    {
+        $this->expectException(ConnectException::class);
+
+        $this->client
+            ->expects($this->any())
+            ->method('request')
+            ->willThrowException(new ConnectException('helloWorld', new Request('post', 'localhost')));
+
+        $this->serverPool->execute('helloWorld', []);
+    }
+
+    public function testExecuteWithFirstFailing()
+    {
+        $body = json_encode(['rows' => [], 'cols' => [], 'duration' => 0, 'rowcount' => 0]);
+
+        $this->client
+            ->expects($this->at(0))
+            ->method('request')
+            ->willThrowException(new ConnectException('helloWorld', new Request('post', 'localhost')));
+
+        $this->client
+            ->expects($this->at(1))
+            ->method('request')
+            ->willReturn(new Response(200, [], $body));
+
+        $this->assertInstanceOf(Collection::class, $this->serverPool->execute('helloWorld', []));
+    }
+
+    public function testWithBadRequest()
+    {
+        $this->expectException(RuntimeException::class);
```

Comment:
Please put the `expectException` call just before the actual call that will trigger it. Too high up here.